demos/mqtt/mqtt_demo_connection_sharing/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,308 @@
+            matchFound = matchWildcards( pTopicFilter,
+                                         pTopicName,
+                                         topicNameLength,
+                                         filterIndex,
+                                         &nameIndex,
+                                         &status );
+        }
+
+        if( matchFound == true )
+        {
+            break;
+        }
+
+        /* Increment indexes. */
+        nameIndex++;
+        filterIndex++;
+    }
+
+    if( status == false )
+    {
+        /* If the end of both strings has been reached, they match. */
+        status = ( ( nameIndex == topicNameLength ) && ( filterIndex == topicFilterLength ) );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool topicMatch( const char * pTopicName,
+                        const uint16_t topicNameLength,
+                        const char * pTopicFilter,
+                        const uint16_t topicFilterLength,
+                        bool requiresExactMatch )
+{
+    bool status = false;
+
+    /* Check for an exact match. */
+    if( topicNameLength == topicFilterLength )
+    {
+        status = ( strncmp( pTopicName, pTopicFilter, topicNameLength ) == 0 );
+    }
+
+    /* If  an exact match is required, return the result of the comparison above.
+     * Otherwise, attempt to match with MQTT wildcards in topic filters. */
+    if( requiresExactMatch == false )
+    {
+        status = topicFilterMatch( pTopicName, topicNameLength, pTopicFilter, topicFilterLength );
+    }
+
+    return status;
+}
+
+/* Iterate in the record list till the first record (from the starting index passed) has a topic match. */
+bool findRecordForTopic( const char * pTopicName,
+                         size_t topicNameLength,
+                         size_t * pCursor )
+{
+    bool matchFound = false;
+
+    do
+    {
+        matchFound = topicMatch( pTopicName,
+                                 topicNameLength,
+                                 callbackRecordList[ *pCursor ].pTopicFilter,
+                                 callbackRecordList[ *pCursor ].topicFilterLength,
+                                 callbackRecordList[ *pCursor ].requiresExactMatch );
+        /* code */
+    } while( ++( *pCursor ) < recordListSize && ( matchFound == false ) );
+
+    return matchFound;
+}
+
+
+/* Design -
+ *   * Common handler can contain logic of processing acks and retries.
+ *   * Only forward PUBLISH message to message dispatcher/handler as individual callbacks
+ *     only want to process incoming PUBLISH message.
+ */
+void SubscriptionManager_DispatchHandler( uint16_t packetIdentifier,
+                                          MQTTPublishInfo_t * pPublishInfo )
+{
+    assert( pPublishInfo != NULL );
+    assert( packetIdentifier != MQTT_PACKET_ID_INVALID );
+
+    size_t recordListCursor = 0u;
+
+    while( findRecordForTopic( pPublishInfo->pTopicName,
```

Comment:
I don't see why `findRecordForTopic` is necessary. Since you're iterating over every entry in the list, why not just use a `for` loop in this function?

Reply:
Good point, removed